File: api/stats.js
export default async function handler(req, res) {
    try {
        // 获取查询参数
        const { username, ...otherParams } = req.query;

        if (!username) {
            return res.status(400).json({ error: 'Username is required' });
        }

        // 构建 GitHub Stats API URL
        const params = new URLSearchParams({
            username,
            show_icons: otherParams.show_icons || 'true',
            hide_rank: otherParams.hide_rank || 'true',
            hide_border: otherParams.hide_border || 'true',
            card_width: otherParams.card_width || '350',
            ...otherParams
        });

        const githubStatsUrl = `https://github-readme-stats.vercel.app/api?${params.toString()}`;

        // 获取原始 SVG
        const response = await fetch(githubStatsUrl);
        if (!response.ok) {
            throw new Error('Failed to fetch GitHub stats');
        }

        const originalSvg = await response.text();

        // 创建带有自定义装饰的新 SVG
        const enhancedSvg = `
<svg width="500" height="200" xmlns="http://www.w3.org/2000/svg">
  <rect
    x="0.5"
    y="0.5"
    rx="4.5"
    height="98%"
    width="499"
    fill="#fffefe"
    stroke="#e4e2e2"
    stroke-opacity="1"
  />
  
  <!-- 嵌入原始 GitHub Stats -->
  <g>
    ${extractSvgContent(originalSvg)}
  </g>
  
  <!-- 自定义装饰路径 -->
  <svg x="400" y="0" width="80" viewBox="0 0 230.7 357.6" fill-opacity="0.2">
    <path d="M1045.4,483.4c40.5-5.7,81.4-2,122.1-3.7,7.4-.1,7.4-.1,7.5-7.6s4.5-14.6,12.6-12.1-.5,17.9,8.6,19.2c12.6.7,25.2,2.6,37.6-2,8.9-3.3,12.9-.5,13.3,9s-1.8,15.3-11.5,11.7c-11.5-3.7-23.1-2.9-34.7-2.6-8.3.3-8.4.6-8.9,8.9s-3.2,11-11.5,9c-11.2-4.4,1.7-18.9-16.5-17.1C1124.9,493.3,1080.5,504.6,1045.4,483.4Z"
      transform="translate(-1045.4 -302.8)">
      <animateTransform
        attributeName="transform"
        attributeType="XML"
        type="translate"
        values="-1045.4,-302.8; -1045.4,-320.8; -1045.4,-310.8; -1045.4,-325.8; -1045.4,-302.8"
        keyTimes="0; 0.25; 0.5; 0.75; 1"
        dur="5s"
        repeatCount="indefinite"/>
    </path>
  </svg>
</svg>`.trim();

        // 返回 SVG
        res.setHeader('Content-Type', 'image/svg+xml');
        res.setHeader('Cache-Control', 's-maxage=3600, stale-while-revalidate');
        res.status(200).send(enhancedSvg);

    } catch (error) {
        console.error('Error:', error);
        res.status(500).json({ error: 'Internal server error' });
    }
}

// 提取 SVG 内容（去除外层 svg 标签）
function extractSvgContent(svgString) {
    const match = svgString.match(/<svg[^>]*>([\s\S]*)<\/svg>/i);
    return match ? match[1] : svgString;
}

File: api/stats.go
package main

import (
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"

	"github.com/labstack/echo/v4"
)

// extractSVGContent returns inner content of the first <svg>...</svg>
func extractSVGContent(svg string) string {
	startTagIdx := strings.Index(strings.ToLower(svg), "<svg")
	if startTagIdx == -1 {
		return svg
	}
	gtIdx := strings.Index(svg[startTagIdx:], ">")
	if gtIdx == -1 {
		return svg
	}
	contentStart := startTagIdx + gtIdx + 1
	endIdx := strings.LastIndex(strings.ToLower(svg), "</svg>")
	if endIdx == -1 || endIdx <= contentStart {
		return svg
	}
	return svg[contentStart:endIdx]
}

var safeNameRe = regexp.MustCompile(`^[A-Za-z0-9_-]+$`)

func sanitizeRole(v string) string {
	if safeNameRe.MatchString(v) {
		return v
	}
	return ""
}

func getAssetsDir() string {
	if v := os.Getenv("ASSETS_DIR"); v != "" {
		return v
	}
	return "assets"
}

func fetchUpstream(base string, q url.Values) (string, error) {
	u, _ := url.Parse(base)
	u.RawQuery = q.Encode()
	resp, err := http.Get(u.String())
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	b, err := io.ReadAll(resp.Body)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

func writeSVGHeadersEcho(c echo.Context) {
	h := c.Response().Header()
	h.Set("Content-Type", "image/svg+xml")
	h.Set("Access-Control-Allow-Origin", "*")
	h.Set("Cache-Control", "s-maxage=3600, stale-while-revalidate")
}

func statsEchoHandler(c echo.Context) error {
	q := c.QueryParams()
	username := strings.TrimSpace(c.QueryParam("username"))
	if username == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "Username is required"})
	}

	// Defaults for outer SVG size
	svgWidth := 500
	svgHeight := 200
	if v := c.QueryParam("svg_width"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			svgWidth = n
		}
	}
	if v := c.QueryParam("svg_height"); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			svgHeight = n
		}
	}

	// Prepare upstream query params (exclude our own custom ones)
	upstreamQ := url.Values{}
	for key, vals := range q {
		if key == "svg_width" || key == "svg_height" || key == "role" || key == "role_x" || key == "role_y" {
			continue
		}
		for _, v := range vals {
			upstreamQ.Add(key, v)
		}
	}
	upstreamQ.Set("username", username)

	// Fetch upstream stats SVG
	orig, err := fetchUpstream("https://github-readme-stats.vercel.app/api", upstreamQ)
	if err != nil {
		return c.JSON(http.StatusBadGateway, map[string]string{"error": "Failed to fetch upstream"})
	}
	leftContent := extractSVGContent(orig)

	// Load role asset if provided
	role := sanitizeRole(c.QueryParam("role"))
	assetContent := ""
	if role != "" {
		assetsDir := getAssetsDir()
		p := filepath.Join(assetsDir, role+".svg")
		if f, err := os.Open(p); err == nil {
			defer f.Close()
			b, err := io.ReadAll(f)
			if err == nil {
				assetContent = extractSVGContent(string(b))
			}
		}
	}

	// Place asset on the right side with a default offset
	assetX := 360
	assetY := 0
	if v := c.QueryParam("role_x"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			assetX = n
		}
	}
	if v := c.QueryParam("role_y"); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			assetY = n
		}
	}

	// Compose final SVG
	var sb strings.Builder
	sb.Grow(len(leftContent) + len(assetContent) + 512)
	sb.WriteString(`<svg width="`)
	sb.WriteString(strconv.Itoa(svgWidth))
	sb.WriteString(`" height="`)
	sb.WriteString(strconv.Itoa(svgHeight))
	sb.WriteString(`" xmlns="http://www.w3.org/2000/svg">`)
	sb.WriteString(`\n  <rect x="0.5" y="0.5" rx="4.5" height="98%" width="`)
	sb.WriteString(strconv.Itoa(svgWidth - 1))
	sb.WriteString(`" fill="#fffefe" stroke="#e4e2e2" stroke-opacity="1"/>`)
	sb.WriteString(`\n  <g>\n`)
	sb.WriteString(leftContent)
	sb.WriteString(`\n  </g>`)
	if assetContent != "" {
		sb.WriteString(`\n  <g transform="translate(`)
		sb.WriteString(strconv.Itoa(assetX))
		sb.WriteString(`,`)
		sb.WriteString(strconv.Itoa(assetY))
		sb.WriteString(`)">\n`)
		sb.WriteString(assetContent)
		sb.WriteString(`\n  </g>`)
	}
	sb.WriteString(`\n</svg>`)

	writeSVGHeadersEcho(c)
	return c.Blob(http.StatusOK, "image/svg+xml", []byte(sb.String()))
}

func registerStats(e *echo.Echo) {
	e.GET("/api/stats", statsEchoHandler)
}
